### PR TITLE
Completed Chapter1!!!

### PR DESCRIPTION
**Pull request by ynfle**
### Solutions for Chapter 1

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter1.hs [hunk 1]
```diff
@@ -541,8 +543,11 @@ value after "=" where the condition is true.
 
 Casual reminder about adding top-level type signatures for all functions :)
 -}
-
-mid x y z = error "mid: not implemented!"
+mid :: Int -> Int -> Int -> Int
+mid x y z
+    | x > y = mid y x z
+    | y > z = mid x z y
+    | otherwise = y
```
Comment (vrom911) [inline src/Chapter1.hs:550]: Very elegant recursive solution with the sorting 😍

File: src/Chapter1.hs [hunk 2]
```diff
@@ -619,8 +631,8 @@ Implement a function that returns the sum of the last two digits of a number.
 Try to introduce variables in this task (either with let-in or where) to avoid
 specifying complex expressions.
 -}
-
-sumLast2 n = error "sumLast2: Not implemented!"
+sumLast2 :: Int -> Int
+sumLast2 n = (lastDigit (abs n)) + (lastDigit (div (abs n) 10))
```
Comment (vrom911) [inline src/Chapter1.hs]: Great one!
Note that you can omit the parenthesis in here:

```suggestion
sumLast2 n = lastDigit (abs n) + lastDigit (div (abs n) 10)
```

Also, as you see that you use `abs n` twice in one function. That means that you could move it into a separate variable in the `let` or `where` clause and reuse everywhere 🙂

File: src/Chapter1.hs [hunk 3]
```diff
@@ -640,8 +652,10 @@ Implement a function that returns the first digit of a given number.
 You need to use recursion in this task. Feel free to return to it later, if you
 aren't ready for this boss yet!
 -}
-
-firstDigit n = error "firstDigit: Not implemented!"
+firstDigit :: Int -> Int
+firstDigit n
+    | abs n <= 9 = abs n
+    | otherwise = firstDigit (div (abs n) 10)
```
Comment (vrom911) [inline src/Chapter1.hs]: Wonderful! 👌🏼 

Here, you can also move `abs n` into a variable in the `where` construction 🙂

**Review by vrom911 — APPROVED** [review]
Neat work! Congrats on finishing Chapter One 🍾 
Very accurate solutions and clear code 👏🏼 

Feel free to merge it as it is 💯  Also note, that there are new Hacktoberfest rules, so don't forget to add the topic to this repo and label to the PR 🙂

File: src/Chapter1.hs (hunk 2 above)
Comment (ynfle) [reply src/Chapter1.hs]: Oh that makes sense

File: src/Chapter1.hs (hunk 1 above)
Comment (ynfle) [reply src/Chapter1.hs:550]: I actually saw it on someone else's solution and was mindblow 🤯

File: src/Chapter1.hs (hunk 3 above)
Comment (ynfle) [reply src/Chapter1.hs]: Would you recommend `where or `let`?
What is considered more idiomatic/what are the benefits?

**Comment by ynfle** [comment]
I have to say that it is tremendously generous of you 2 for doing this!!! Thanks a ton!

File: src/Chapter1.hs (hunk 3 above)
Comment (vrom911) [reply src/Chapter1.hs]: I personally prefer `where` as it allows me to first look at the wider picture, only then on the details in `where`.

`where` and `let` have different scoping boundaries, so you need to be careful and use the appropriate constructions.
In here, you can not write `let` before `guards`, so you would need to add it in the case itself, but that would mean that it works only inside that scope and won't be visible in another guard case.
`where` on the other hand, would be visible in scopes of both cases. 🙂 

This is not an easy topic to explain all details of the difference between `let` and `where`, so we decided not to include it in the course itself. But I hope the above makes some sense. You can ask any questions and we could discuss these constructions and scoping in details if you are interested 👌🏼

Comment (ynfle) [reply src/Chapter1.hs]: I ended up using both. `where` for this one and `let` for the other.

Comment (ynfle) [reply src/Chapter1.hs]: That is very interesting that `let` would work on a guard-per-guard basis

**Pull request merged**